nntrainer/graph/graph_node.h

```diff
@@ -74,5 +75,228 @@ class GraphNode {
   // virtual GraphNode &copy(const GraphNode &from) = 0;
 };
 
+/**
+ * @brief   Iterator for GraphNode which return LayerNode object upon realize
+ *
+ * @note    This does not include the complete list of required functions. Add
+ * them as per need.
+ */
+template <typename T>
+class GraphNodeIterator : public std::iterator<std::random_access_iterator_tag,
+                                               std::shared_ptr<GraphNode>> {
+  std::shared_ptr<GraphNode> *p; /** underlying object of GraphNode */
+
+public:
+  /**
+   * @brief   iterator_traits types definition
+   *
+   * @note    these are not requried to be explicitly defined now, but maintains
+   *          forward compatibility for c++17 and later
+   *
+   * @note    valute_type, pointer and reference are different from standard
+   * iterator
+   */
+  typedef std::shared_ptr<T> value_type;
+  typedef std::random_access_iterator_tag iterator_category;
```

Comment:
Is it allowed and intended to allow random access?

Reply:
I guess we could stay Bidirectional iterator if it is yet to be implemented.

Reply:
Given the underlying data structure is vector, I kept random_access.
Also, we provide getNodeByIdx()/getNodeByName() which are also providing random_access.

Reply:
Okay